src/inmanta/server/services/orchestrationservice.py

```diff
@@ -611,72 +576,85 @@ def safe_get(input: object, key: str, default: object) -> object:
                     resource_version_id=ResourceVersionIdStr("%s,v=%s" % (res.resource_id, version)),
                     attributes=attributes,
                 )
-                resource_objects.append(res_obj)
 
                 previous_requires[res_obj.resource_id] = res.attributes["requires"]
-                resource_version_ids.append(res_obj.resource_version_id)
-                agents.add(res_obj.agent)
-                rv_dict[res_obj.resource_id] = res_obj
+                rid_to_resource[res_obj.resource_id] = res_obj
 
             # invert dependencies on purges
```

Comment:
We do it because most often, deletes have to be done in reverse order (i.e. create folder, then file, delete file, then folder)

Reply:
I looked it over once more and I draw the following conclusions:
1. My earlier statement about inverting the requires relation twice doesn't hold.
2. Suppose `a` requires `b`, where `a` and `b` live on different agents and only `a` is deleted from the model. Then this logic goes through the following steps
  a. empty `a`'s requires and provides
  b. leave `b`'s requires and provides as is because it is not being purged
  c. invert: set `b.requires = a` and `a.provides = b`
  => `b` still has `b.provides = a`, resulting in a circular (and incorrect) dependency. A similar issue occurs if the dependency would have been the other way around.

Do you agree (2) is a bug? I don't think it's introduced by this PR.

Reply:
looks like a bug to me indeed

Reply:
I agree this is a bug.